File: src/safe_ride_flutter/lib/features/customer/booking/data/models/create_booking_request.dart
import '../../../../../core/constants/app_strings.dart';
import 'booking_location.dart';

enum BookingType { now, scheduled }

class CreateBookingRequest {
  const CreateBookingRequest({
    required this.vehicleId,
    required this.serviceTypeId,
    required this.bookingType,
    required this.pickup,
    this.destination,
    this.scheduledAt,
    this.specialRequest,
    this.estimatedHours,
    this.promotionCode,
  });

  final int vehicleId;
  final int serviceTypeId;
  final BookingType bookingType;
  final DateTime? scheduledAt;
  final BookingLocation pickup;
  final BookingLocation? destination;
  final String? specialRequest;
  final int? estimatedHours;
  final String? promotionCode;

  CreateBookingRequest copyWith({
    int? vehicleId,
    int? serviceTypeId,
    BookingType? bookingType,
    DateTime? scheduledAt,
    BookingLocation? pickup,
    BookingLocation? destination,
    String? specialRequest,
    int? estimatedHours,
    String? promotionCode,
  }) {
    return CreateBookingRequest(
      vehicleId: vehicleId ?? this.vehicleId,
      serviceTypeId: serviceTypeId ?? this.serviceTypeId,
      bookingType: bookingType ?? this.bookingType,
      scheduledAt: scheduledAt ?? this.scheduledAt,
      pickup: pickup ?? this.pickup,
      destination: destination ?? this.destination,
      specialRequest: specialRequest ?? this.specialRequest,
      estimatedHours: estimatedHours ?? this.estimatedHours,
      promotionCode: promotionCode ?? this.promotionCode,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      ApiKeys.vehicleId: vehicleId,
      ApiKeys.serviceTypeId: serviceTypeId,
      ApiKeys.bookingType: bookingType == BookingType.now
          ? AppValues.bookingNow
          : AppValues.bookingScheduled,
      ApiKeys.scheduledAt: scheduledAt?.toUtc().toIso8601String(),
      ApiKeys.pickupAddress: pickup.address,
      ApiKeys.pickupLatitude: pickup.latitude,
      ApiKeys.pickupLongitude: pickup.longitude,
      ApiKeys.destinationAddress: destination?.address,
      ApiKeys.destinationLatitude: destination?.latitude ?? pickup.latitude,
      ApiKeys.destinationLongitude: destination?.longitude ?? pickup.longitude,
      ApiKeys.specialRequest: specialRequest?.trim().isEmpty == true
          ? null
          : specialRequest?.trim(),
      ApiKeys.estimatedHours: estimatedHours,
      if (promotionCode != null && promotionCode!.trim().isNotEmpty)
        ApiKeys.promotionCode: promotionCode!.trim(),
    };
  }
}

File: src/SRD_Master/SafeRide.Contracts/Requests/Bookings/CreateBookingRequest.cs
using System.ComponentModel.DataAnnotations;
using SafeRide.Domain.Enums;

namespace SafeRide.Contracts.Requests.Bookings;

public sealed class CreateBookingRequest
{
    public BookingType BookingType { get; init; }
    public DateTime? ScheduledAt { get; init; }

    [Range(1, long.MaxValue)]
    public long VehicleId { get; init; }

    [Range(1, long.MaxValue)]
    public long ServiceTypeId { get; init; }

    [Required, MaxLength(255)]
    public string PickupAddress { get; init; } = string.Empty;

    [Range(-90, 90)]
    public double PickupLatitude { get; init; }

    [Range(-180, 180)]
    public double PickupLongitude { get; init; }

    [MaxLength(255)]
    public string? DestinationAddress { get; init; }

    [Range(-90, 90)]
    public double DestinationLatitude { get; init; }

    [Range(-180, 180)]
    public double DestinationLongitude { get; init; }

    [MaxLength(500)]
    public string? SpecialRequest { get; init; }

    [Range(1, 24)]
    public int? EstimatedHours { get; init; }

    [MaxLength(50)]
    public string? PromotionCode { get; init; }
}
